Added a `Linear` TS search job adapter

alongd opened the pull request:
Adds a new Linear TS search adapter that builds TS guesses from atom-mapped reactants and products via Z-matrix chimera construction with Hammond-biased weighting. The adapter is incore-only, plugs into ARC's scheduler like heuristics/autotst_ts, and delegates heavy geometry work to a new linear_utils/ subpackage (5 modules). Currently implemented for isomerization/unimolecular reactions.

The PR also carries supporting additions to arc/species/zmat.py (anchors, smart-anchor detection, zmat re-indexing helpers), arc/species/converter.py (atom-map reordering), arc/reaction/reaction.py (is_unimolecular, refined is_isomerization), and a biradical-preservation fix in arc/species/species.py. CI was hardened: pinned action, `-n 4 --dist worksteal` for stability, and obabel test made self-contained.

alongd committed:
Added Reaction.is_unimolecular()

Also fixed the is_isomerization() method

alongd committed:
Added Reaction.is_unimolecular()

Also fixed the is_isomerization() method

alongd committed:
Added an atom_order arg to xyz_to_zmat()

alongd committed:
Add atom_order and anchors arguments to xyz_to_zmat()

Combines the previously separate atom_order and anchors extensions to xyz_to_zmat(), along with their tests.

alongd committed:
Tests: xyz_to_zmat()

Also testing the atom_order arg

alongd committed:
Added 'linear' to ts_adapters_by_rmg_family

alongd committed:
Added 'linear' to ts_adapters_by_rmg_family

alongd committed:
Added check_ordered_zmats()

alongd committed:
Added check_ordered_zmats()

alongd committed:
Tests: check_ordered_zmats()

alongd committed:
Added linear to JobEnum

alongd committed:
Added linear to JobEnum

alongd committed:
Tests: Reaction.is_unimolecular()

alongd committed:
Tests: Reaction.is_unimolecular()

alongd committed:
Added the Linear TS search job adapter and related utils

alongd committed:
Added the Linear TS search job adapter and related utils

alongd committed:
Tests: Linear TS Job Adapter and utils

alongd committed:
Tests: Linear TS Job Adapter and utils

alongd committed:
Added order_xyz_by_atom_map to converter

alongd committed:
Added order_xyz_by_atom_map to converter

And tests

alongd committed:
Tests: converter order_xyz_by_atom_map()

alongd committed:
Added update_zmat_by_xyz() to zmat

alongd committed:
Added update_zmat_by_xyz() to zmat

alongd committed:
Tests: zmat update_zmat_by_xyz()

alongd committed:
Use the round_to arg in common get_angle_in_180_range()

alongd committed:
Use the round_to arg in common get_angle_in_180_range()

And tests

alongd committed:
Tests: use round_to in get_angle_in_180_range()

alongd committed:
Added anchors to xyz_to_zmat()

And updated update_zmat_by_xyz()

alongd committed:
Tests: zmat anchors

alongd committed:
Minor: Style mod in mapping driver

alongd committed:
Remove Julia installation from install_rmg

alongd committed:
Remove Julia installation from install_rmg

and use run_devtool in install_all for pyrdl

alongd committed:
Added converter order_mol_by_atom_map()

And tests

alongd committed:
Added converter order_mol_by_atom_map()

And tests

alongd committed:
Added zmat find_smart_anchors()

alongd committed:
Added zmat find_smart_anchors()

alongd committed:
Updated the ts_adapters_by_rmg_family dict

added the linear adapter families

alongd committed:
Updated the ts_adapters_by_rmg_family dict

added the linear adapter families

alongd committed:
Preserve radical sites when perceiving mol from xyz with an existing mol

When an ARCSpecies is constructed with both an adjlist (or SMILES) and xyz, mol_from_xyz re-perceives the molecule from the 3D geometry. Previously, perceive_molecule_from_xyz was called with n_radicals=None (the default) even when the existing mol carried radical information. For singlet biradicals this caused the perception to produce a closed-shell molecule with lone pairs instead of radical centers, which then replaced the original mol and broke downstream family detection (e.g.,  Intra_Disproportionation was not identified).

Now, when self.mol already has radical electrons and the user did not explicitly set number_of_radicals, the radical count is derived from self.mol and forwarded to perceive_molecule_from_xyz and is_mol_valid.

alongd committed:
Preserve radical sites when perceiving mol from xyz with an existing mol

When an ARCSpecies is constructed with both an adjlist (or SMILES) and xyz, mol_from_xyz re-perceives the molecule from the 3D geometry. Previously, perceive_molecule_from_xyz was called with n_radicals=None (the default) even when the existing mol carried radical information. For singlet biradicals this caused the perception to produce a closed-shell molecule with lone pairs instead of radical centers, which then replaced the original mol and broke downstream family detection (e.g.,  Intra_Disproportionation was not identified).

Now, when self.mol already has radical electrons and the user did not explicitly set number_of_radicals, the radical count is derived from self.mol and forwarded to perceive_molecule_from_xyz and is_mol_valid.

alongd committed:
Tests: call self.job_3.execute() in OB tests

So that parallel execution of tests on CI won't run w/o job.execute()

alongd committed:
Tests: call self.job_3.execute() in OB tests

So that parallel execution of tests on CI won't run w/o job.execute()

alongd committed:
Modifications to CI

Added 4 workers to speedup (auto gave 2)

alongd committed:
Modifications to CI

Added 6 workers to speedup (auto gave 2)

alongd committed:
Added an iPY notebook to showcase the linear adapter

alongd committed:
Added an iPY notebook to showcase the linear adapter

alongd committed:
Docs: Added descriptions of TS search methods implemented in ARC

Also added an RMG installation note

alongd committed:
Docs: Added descriptions of TS search methods implemented in ARC

alongd committed:
Added split_entries() to family

To read additional types of RMG-database families

alongd committed:
Added split_entries() to family

To read additional types of RMG-database families

alongd committed:
Organize teardown in functional and common tests

alongd committed:
Organize teardown in functional and common tests

alongd committed:
Added a debug message if fragment mapping fails

and return None

alongd committed:
Added a debug message if fragment mapping fails

and return None

alongd committed:
Added  .coverage* and Claude to .gitignore

alongd committed:
Added  Claude to .gitignore

alongd committed:
Minor: Style modifications

UK -> US English spelling, remove excessive hyphenation

alongd committed:
Minor: Style modifications

UK -> US English spelling, remove excessive hyphenation

alongd committed:
Tests: Added family tests

alongd committed:
Tests: Added family tests

alongd committed:
Implement ts_adapters_for_unknown_unimolecular() in Scheduler

alongd committed:
Implement ts_adapters_for_unknown_unimolecular() in Scheduler

alongd committed:
Added economic_generation to conformers

so scissors / atom mapping / tests can use fewer confs

alongd committed:
Added economic_generation to conformers

so scissors / atom mapping / tests can use fewer confs

alongd committed:
Tests: economic conformer generation

alongd committed:
Tests: economic conformer generation

alongd committed:
Use economic_generation in scissors

alongd committed:
Use economic_generation in scissors

alongd committed:
Minor: Added exist_ok to project directory os.makedirs in main

alongd committed:
Minor: Added exist_ok to project directory os.makedirs in main

alongd committed:
Tests: Extended perception test

To allow another SMILES string.
Actually, the test isn't great, the multiplicity cannot be 1, it should probably be 2. However, multiplicity 2 gives an even worse result with many radicals: 'OCCC(C(COO)(O[CH2])C)(C[C]1[CH][CH][C]([CH][CH]1)NO)SC' (all carbon atoms on the benzene ring are marked as having a radical).
This is out of scope for the current linear TS adapter PR. The fix should be easy (detect a possibly aromatic ring with all  radicals, at least for the C-only case, then make it aromatic). Leaving it as is for now since this is indeed a crazy molecule.

alongd committed:
Tests: Extended perception test

To allow another SMILES string.
Actually, the test isn't great, the multiplicity cannot be 1, it should probably be 2. However, multiplicity 2 gives an even worse result with many radicals: 'OCCC(C(COO)(O[CH2])C)(C[C]1[CH][CH][C]([CH][CH]1)NO)SC' (all carbon atoms on the benzene ring are marked as having a radical).
This is out of scope for the current linear TS adapter PR. The fix should be easy (detect a possibly aromatic ring with all  radicals, at least for the C-only case, then make it aromatic). Leaving it as is for now since this is indeed a crazy molecule.

alongd committed:
cache ReactionFamily instances and groups.py reads per process

In arc/family/family.py:

  1. Caches ReactionFamily instances per process. Adds a _cache: Dict[(label, consider_arc_families), ReactionFamily] class dict, plus a __new__ that  returns the cached instance for a repeat key. __init__ early-returns on cache hits via a _initialized flag, so the parsing work runs at most once per
   family per process.
  2. Caches groups.py file reads. Extracts the file-loading logic out of ReactionFamily.get_groups_file_as_lines into a module-level _read_groups_file_lines(label, consider_arc_families) decorated with @functools.lru_cache(maxsize=None). Returns a tuple (immutable) so callers can't poison the cache. The method now just delegates.
  3. Caches recommended family sets. Decorates get_rmg_recommended_family_sets with @functools.lru_cache(maxsize=1) since its result is also process-stable.
  4. Moves the label is None guard from __init__ to __new__ (it has to run before cache lookup).

Net effect: repeated ReactionFamily(label) calls — common in mapping/heuristics — stop re-reading and re-parsing the same groups.py.

alongd committed:
cache ReactionFamily instances and groups.py reads per process

In arc/family/family.py:

  1. Caches ReactionFamily instances per process. Adds a _cache: Dict[(label, consider_arc_families), ReactionFamily] class dict, plus a __new__ that  returns the cached instance for a repeat key. __init__ early-returns on cache hits via a _initialized flag, so the parsing work runs at most once per
   family per process.
  2. Caches groups.py file reads. Extracts the file-loading logic out of ReactionFamily.get_groups_file_as_lines into a module-level _read_groups_file_lines(label, consider_arc_families) decorated with @functools.lru_cache(maxsize=None). Returns a tuple (immutable) so callers can't poison the cache. The method now just delegates.
  3. Caches recommended family sets. Decorates get_rmg_recommended_family_sets with @functools.lru_cache(maxsize=1) since its result is also process-stable.
  4. Moves the label is None guard from __init__ to __new__ (it has to run before cache lookup).

Net effect: repeated ReactionFamily(label) calls — common in mapping/heuristics — stop re-reading and re-parsing the same groups.py.

alongd committed:
Add pytest configuration with parallel execution and test paths

alongd committed:
Add pytest configuration with parallel execution and test paths

calvinp0 commented on arc/job/adapters/ts/linear_utils/addition.py:
```diff
@@ -0,0 +1,1524 @@
+            frags.append(component)
+        return frags
+
+    def _heavy_formulas(formulas: List[Dict[str, int]]) -> List[Dict[str, int]]:
```
I personally am not the biggest fan of nested functions - are we able to pull these out? I feel it helps with readability cause I got a bit confused at like 92/93 thinking these were two floating variables between functions before realising this was all in one function

calvinp0 commented on arc/job/adapters/ts/linear_utils/addition.py:
```diff
@@ -0,0 +1,1247 @@
+        try:
+            mol_copy.get_bond(copy_atoms[a_idx], copy_atoms[b_idx])
+            continue  # bond already exists
+        except (ValueError, KeyError):
```
Yeah, I think we should not do bare excepts

alongd replied:
agree, fixed as suggested

alongd committed:
Type hints leftover fixes

calvinp0 commented on .github/workflows/ci.yml:
```diff
@@ -23,21 +23,21 @@ jobs:
           environment-name: arc_env
           environment-file: ARC/environment.yml
-          cache-environment: true
+          cache-environment: false
```
is there a reason we are not caching the environment?

alongd replied:
To keep the nightly cron (line 9 of ci) catch env drift issues, I'll implement:
`cache-environment: ${{ github.event_name != 'schedule' }}` to make PR's CI faster

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    # Calibrated TS target distances for [1,3]-sigmatropic shifts.
+    sbl_break = get_single_bond_length(symbols[origin], symbols[migrating]) or 1.5
+    sbl_form = get_single_bond_length(symbols[target], symbols[migrating]) or 1.5
+    _SIGMA_BREAK_STRETCH = 0.77
```
I think these (incl. _SIGMA_FORM_STRETCH) should become module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+            if symbols[ib] == 'H':
+                continue
+            d = float(np.linalg.norm(coords[ia] - coords[ib]))
+            if d < 0.9:
```
Is there also a situation where it can be wildly above?

alongd replied:
In principle, yes. Currently we catch collisions, I'll also add a detection for very large distances. Thanks!

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    Given two spheres of radii ``r1`` (around ``center1``) and ``r2``
+    (around ``center2``), this returns the intersection point that lies
+    on the same side of the inter-center axis as ``ref_pos``. When the
+    spheres do not overlap, the point is placed at distance ``r1`` from
```
Maybe its just worth mentionig in the doctstring but if there is a case when th circle is inside the other circle completely. I guess it depends if you consider that as a 'no-intersection' and if that situation is every possible.

alongd replied:
I agree. Currently it fails if one circle is within the other. I'll add a fallback to collinear placement

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+            o_hydroxyl = nbr_idx
+            break
+
+    _BV_CO_SHORTEN_TARGET = 1.28  # Å, calibrated from DFT TS (1.279)
```
Module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    sbl_cc = get_single_bond_length('C', 'C') or 1.54
+    sbl_co = get_single_bond_length('C', 'O') or 1.43
+
+    _BV_OO_STRETCH = 0.27    # Å per side (total ≈ 2 × 0.27 above sbl)
```
Module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    sbl_co = get_single_bond_length('C', 'O') or 1.43
+
+    _BV_OO_STRETCH = 0.27    # Å per side (total ≈ 2 × 0.27 above sbl)
+    _BV_CC_STRETCH = 0.76    # Å above sbl for migrating-group departure
```
Module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+
+    _BV_OO_STRETCH = 0.27    # Å per side (total ≈ 2 × 0.27 above sbl)
+    _BV_CC_STRETCH = 0.76    # Å above sbl for migrating-group departure
+    _BV_CO_STRETCH = 0.73    # Å above sbl(C,O) for migrating-group approach
```
Module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+
+        # 4d. Concerted H migration from O_hydroxyl to the carbonyl O
+        # (~1.40 Å from each, calibrated).
+        _BV_H_TRANSFER_TARGET = 1.40
```
Module level

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+                h_idx = h_on_oh[0]
+                p1 = coords[o_hydroxyl]
+                p2 = coords[o_carbonyl_dbl]
+                ax = p2 - p1
```
Is this not something already kinda done in _two_sphere_intersection function?

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+        frag_mig = bfs_fragment(adj, c_migrating, block={c_parent})
+        p1 = coords[c_parent]
+        p2 = coords[o_other_side]
+        axis = p2 - p1
```
Is this not something similar in _two_sphere_intersection?

alongd replied:
you're right, the math is duplicated. also, I see it in other families.py locations and in local_geometry, addition, and postprocess. I'll consolidate

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    for c_migrating in mig_candidates:
+        coords = np.array(uni_xyz['coords'], dtype=float).copy()
+
+        # 4a. Stretch the O-O peroxide bond.
```
It's under the comment of step 5 but then here its 4a (then 4b etc.)

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+        d_oo = float(np.linalg.norm(coords[oo_bond[0]] - coords[oo_bond[1]]))
+        d_cc = float(np.linalg.norm(coords[c_parent] - coords[c_migrating]))
+        d_co = float(np.linalg.norm(coords[c_migrating] - coords[o_other_side]))
+        score = ((d_oo - d_oo_target) ** 2
```
if the function also does C_parent–O_hydroxyl shortening and H transfer, the score ignores those by the looks of it. Maybe do add optionl score terms when those atoms exist?
Cause if two candidates tie on the 3 main distances, one could, I guess, have a muc better or worse H migration geom

alongd replied:
the concern is legit, but the impact might be nuanced in my opinion.
I added two optional terms to the score, gated on `o_hydroxyl is not None` and `o_carbonyl_dbl is not None`

calvinp0 commented on arc/job/adapters/ts/linear_utils/families.py:
```diff
@@ -0,0 +1,1596 @@
+    # Step 1: Find the O-O peroxide bond.
+    oo_bond = None
+    for a, b in split_bonds:
+        if symbols[a] == 'O' and symbols[b] == 'O':
```
will there ever be a chance when a/b can be not valid indexes?

alongd replied:
Thanks, I appreciate the question. Here, `a` and `b` are always valid indices
Of course, if we have nonsense input the function could break, but this is also true for other params in this function, and to many others. I don't think we should guard against all possibilities, the current state is realistic, I'll leave as-is

calvinp0 commented on arc/family/family.py:
```diff
@@ -40,10 +41,39 @@ def get_rmg_db_subpath(*parts: str, must_exist: bool = False) -> str:
 
+    Instances are cached per ``(label, consider_arc_families)`` so the
+    family ``groups.py`` file is read and parsed at most once per process.
+    The cached object is treated as immutable; do not mutate its public
```
Just to double check, are any of `self.reactants`, `self.entries`, or `self.actions` ever mutated downstream? If so, then the cache can leak state between calls

alongd replied:
no, there's no modification downstream, so no leak risk in the current code. But we're not guarded against future code changes that might introduce mutations. If we want, we could make these tuples instead of lists. I'll leave it as is, I think the risk is low

calvinp0 commented on arc/family/family.py:
```diff
@@ -40,10 +41,39 @@ def get_rmg_db_subpath(*parts: str, must_exist: bool = False) -> str:
 
 
+@functools.lru_cache(maxsize=None)
+def _read_groups_file_lines(label: str, consider_arc_families: bool) -> tuple[str, ...]:
```
So this feeds into `get_groups_file_as_lines` right, which method says def get_groups_file_as_lines(...) -> list[str]:

calvinp0 replied:
Also since we have this function, it seems redundant to then have a self function inside the class thats sole purpose is to call this function? But okay with leaving it as is, just thought to point it out

alongd replied:
Thanks, I'll make the helper public and drop the wrapper

calvinp0 commented on arc/family/family.py:
```diff
@@ -40,10 +41,39 @@ def get_rmg_db_subpath(*parts: str, must_exist: bool = False) -> str:
+        return tuple(f.readlines())
+
+
 class ReactionFamily(object):
```
Since ReactionFamily is now cached, there is a line that appears in nested loops
```
Group().from_adjacency_list(get_group_adjlist(...))
```

We should consider caching it too?

```
self.groups_by_label = {
    label: Group().from_adjacency_list(adjlist)
    for label, adjlist in self.entries.items()
}
```

alongd replied:
Good catch! applied

calvinp0 commented:
Not in this PR, but it does touch this file significantly `family.py`
https://github.com/ReactionMechanismGenerator/ARC/blob/58770524bab03e3cfe649875ec5b2a37a91c6968/arc/family/family.py#L713
I believe that two lines above, we just initialised `rmg_families, arc_families = list(), list()`, so it is always `False`

calvinp0 commented on .github/workflows/ci.yml:
```diff
@@ -23,21 +23,21 @@ jobs:
           dotnet: true
           haskell: true
           large-packages: true
-          swap-storage: true
```
Were we getting OOM when this was true?

alongd replied:
We weren't getting OOMs, but moving from 2 to 6 workers in CI risked OOM under peak load

alongd commented:
@calvinp0, comments addressed, ready for the next round. Thanks!

calvinp0 commented on arc/job/adapters/ts/linear_utils/math_zmat.py:
```diff
@@ -0,0 +1,388 @@
+
+    if reorg_energy is None:
+        lam_endo = delta_e_sat / (2.0 * (w_max - 0.5)) if (w_max - 0.5) > 0 else float('inf')
+        lam_exo  = delta_e_sat / (2.0 * (0.5 - w_min)) if (0.5 - w_min) > 0 else float('inf')
```
double spacing for `lam_exo  =`

alongd replied:
this is "on purpose" for comparability of the two lines, so it's easier to see `(w_max - 0.5)` directly above `(0.5 - w_min)`

calvinp0 commented on arc/job/adapters/ts/linear_utils/math_zmat.py:
```diff
@@ -0,0 +1,388 @@
+                                                       key=lambda item: (-item[1], item[0]))]
+    for i, atom in enumerate(atoms_sorted_by_frequency):
+        for bond in expected_breaking_bonds + expected_forming_bonds:
+            if atom in bond and all(a not in atoms_sorted_by_frequency[:i] for a in bond):
```
I want to double-check the intended behavior here. `get_r_constraints()` does not include all changing bonds. For each atom, it picks the first eligible bond and then `break`s. Also, because of the `all(a not in atoms_sorted_by_frequency[:i] for a in bond)` condition, bonds involving an earlier higher-frequency atom are skipped for later atoms.

So for changing bonds like `[(1, 2), (1, 3)]`, this would only keep one constraint involving atom 1, not both. Is the goal to produce a minimal canonical subset of R constraints, or should every expected forming/breaking bond appear in `R_atom`?

alongd replied:
Intentional, it's a minimal canonical R-constraint set: one constraint per reactive atom, not one per bond. An atom appearing in multiple bonds (e.g. a migrating H in both a break and a form) contributes a single R, which avoids over-constraining the optimizer. Added a docstring note to clarify.

calvinp0 commented on arc/job/adapters/ts/linear_utils/path_spec.py:
```diff
@@ -0,0 +1,1461 @@
+    Conservative checks:
+
+    1. **Failed to break** — any bond in ``path_spec.breaking_bonds`` whose
+       distance is below 1.30 Å.
```
Is this true for all bonds? Or would we possible scale it using the single bond length as the base

alongd replied:
Thanks, I agree it'll be more chemically correct to scale it. Adding a scale using `common.get_single_bond_length()`

calvinp0 commented on arc/job/adapters/ts/linear_utils/isomerization.py:
```diff
@@ -0,0 +1,1248 @@
+    coords = np.array(r_xyz['coords'], dtype=float)
+    atom_to_idx = atom_index_map(r_mol)
+
+    scale = 2.0 * (1.0 - weight)
```
I just want to double check, so if weight is 1.0 then scale is 0 and then we expect the PAULING_DELTA to have no impact?

alongd replied:
correct. weight=1.0 means "geometry is fully product-like"

calvinp0 commented on arc/job/adapters/ts/linear_utils/local_geometry.py:
```diff
@@ -0,0 +1,1157 @@
+# ---------------------------------------------------------------------------
+
+def clean_migrating_h(xyz: dict,
+                      mol: Molecule,
```
is `mol` used in this func? I don't see it being called

alongd replied:
removing `mol` and `weight`. thanks

calvinp0 commented on arc/job/adapters/ts/linear_utils/postprocess.py:
```diff
@@ -0,0 +1,2181 @@
+    return xyz_with_coords(xyz, coords)
+
+
+def orient_h_on_reactive_centers(xyz: dict,
```
Is this not kinda the same as https://github.com/ReactionMechanismGenerator/ARC/blob/ea9392510f2b1a14508b4716eb9a043cccba809b/arc/job/adapters/ts/linear_utils/local_geometry.py#L187

I think a lot of funcs here are similar in local_geometry.py
Is there a way to consolidate these functions in a file?

alongd replied:
these two have different underlying logic. I'll check if there are additional funcs to consolidate (already did a pass in the last iteration, I'll check this file specifically now)

calvinp0 commented on arc/job/adapters/ts/linear_utils/postprocess.py:
```diff
@@ -0,0 +1,2181 @@
+        d_target = sbl + PAULING_DELTA
+        # Only act when interpolation failed to stretch at all; do not touch
+        # already-stretched bonds from well-interpolated TSs.
+        if d_cur < sbl:
```
adjust_reactive_bond_distances() docstring says breaking heavy-atom bonds shorter than their TS estimate are stretched, but the code only stretches when the current distance is shorter than the single-bond length
The inline comment seems to explain this, but maybe should be in the docstring too

alongd replied:
I agree, fixed

calvinp0 commented on arc/job/adapters/ts/linear.py:
```diff
@@ -0,0 +1,4030 @@
+# ---------------------------------------------------------------------------
+
+def _frag_composition_key(uni_mol: Molecule,
+                          atom_to_idx: dict,
```
I don't see `atom_to_idx` used in this helper

alongd replied:
fixed

calvinp0 commented on arc/job/adapters/ts/linear.py:
```diff
@@ -0,0 +1,4030 @@
+                frag_centers.add(a_loc)
+                frag_exempt_hs.add(h_idx_loc)
+            if frag_centers:
+                ts_xyz = apply_reactive_center_cleanup(
```
Refer to prev comment

calvinp0 commented on arc/job/adapters/ts/linear.py:
```diff
@@ -0,0 +1,4030 @@
+                        template_exempt_hs.add(int(mig_rec['h_idx']))
+                    if template_centers:
+                        for mig_rec in migration_records:
+                            ts_xyz = orient_h_away_from_axis(
```
This post-migration cleanup pattern appears in both here and then a roughl 400-500 lines down

In both cases, the migrating H has already been placed upstream by `migrate_verified_atoms()` / `migrate_h_between_fragments()`, so we intentionally call:

```python
apply_reactive_center_cleanup(..., migrations=None, exempt_h_indices=...)
```
to avoid calling clean_migrating_h() a second time.

Could we wrap this sequence into a helper, maybe _cleanup_after_existing_h_migration(...)? That would make the “migrating H is already placed; only orient/regularize/symmetrise the donor-acceptor shells” behaviour explicit and avoid the two copies drifting apart later.

alongd replied:
agree, consolidating

calvinp0 commented on arc/species/converter.py:
```diff
@@ -2470,3 +2470,67 @@ def kabsch(xyz1: dict, xyz2: dict) -> float:
     return score
+
+
+def order_xyz_by_atom_map(xyz: dict,
```
See comment  on order_mol_by_atom_map()

calvinp0 commented on arc/species/converter.py:
```diff
@@ -2470,3 +2470,67 @@ def kabsch(xyz1: dict, xyz2: dict) -> float:
+    return xyz_from_data(coords=coords, symbols=symbols, isotopes=isotopes)
+
+
+def order_mol_by_atom_map(mol: Molecule,
```
The two implementations seem internally consistent, but I think the docstring describes the opposite map direction.

The code uses `atom_map` as `new[i] = old[atom_map[i]]`, so `atom_map[new_i] = old_i`. But the docstring says `atom_map[reactant_i] = product_i`, which sounds like old -> new.

Could we clarify the docstring or invert the map if old -> new is the intended convention?

alongd replied:
I extended the docstring to explain this better

calvinp0 commented on arc/species/converter.py:
```diff
@@ -2470,3 +2470,67 @@ def kabsch(xyz1: dict, xyz2: dict) -> float:
+        Molecule: A deep copy of ``mol`` with atoms reordered according to ``atom_map``.
+    """
+    n = len(mol.atoms)
+    if len(atom_map) != n:
```
I think also we should check duplication.

So, for example, [0, 0, 1] passes range/length checks and silently duplicates one atom while dropping another.

alongd replied:
I agree. we check it now with `if len(set(atom_map)) != n`

calvinp0 commented on arc/species/species.py:
Is there a way we can avoid this cyclic import. I want to avoid more cyclic check failures from CodeQL

alongd replied:
converter only imports from species.py within `TYPE_CHECKING`. CodeQL thinks it's a real import. There's no real issue. We can think of options to change the code to silence CodeQL, but I think this is a CodeQL issue, they need to recognize that there's no cyclic import in such cases

calvinp0 commented on arc/species/zmat.py:
```diff
@@ -1324,6 +1353,152 @@ def get_connectivity(mol: Molecule) -> dict[int, list[int]]:
     return connectivity
 
 
+def find_smart_anchors(mol: Molecule,
```
This helper is mostly clear, but the docstring promises “1–3 unique, contiguous atom indices.” The code only excludes the immediately previous atom when picking the next anchor. Could we make that invariant explicit by tracking `used = {anchor_0, anchor_1}` when selecting `anchor_2`, or by validating before return?

alongd replied:
Done. `_pick_next` now takes `used: set[int]` instead of `previous_atom: int`, so every already-chosen anchor is passed in and excluded by construction (rather than relying on graph properties + argument ordering)

calvinp0 commented on arc/species/zmat.py:
```diff
@@ -1324,6 +1353,152 @@ def get_connectivity(mol: Molecule) -> dict[int, list[int]]:
+    spectator_atoms: set = set(range(num_atoms)) - reactive_core
+    connectivity: dict[int, list[int]] = get_connectivity(mol)
+
+    def _atomic_number(atom_index: int) -> int:
```
I see we have nested helper functions here. I will leave it up to you if you want to keep them inside as they are relatively small so not too bad to notice and follow

calvinp0 commented on arc/species/zmat.py:
```diff
@@ -2338,3 +2513,80 @@ def map_index_to_int(index: int | str) -> int:
+    }
+
+    coords = xyz['coords'] if isinstance(xyz, dict) and 'coords' in xyz else xyz
+    n_atoms_xyz = len(coords)
```
`update_zmat_by_xyz()` normalizes `coords` from `xyz`, but then passes `xyz` rather than `coords` into `vectors.calculate_param()`.

alongd replied:
`calculate_param` accepts both forms, but it does seem inconsistent. changes now to `coords`

calvinp0 commented on arc/common_test.py:
```diff
@@ -30,11 +30,33 @@ class TestCommon(unittest.TestCase):
+        if os.path.isfile(globalized_restart_path):
+            try:
+                os.remove(path=globalized_restart_path)
+            except OSError:
```
Should probably log something here if we hit the except statement

calvinp0 commented on functional/functional_test.py:
```diff
@@ -22,14 +22,37 @@ class TestFunctional(unittest.TestCase):
+            if os.path.isfile(stray_path):
+                try:
+                    os.remove(stray_path)
+                except OSError:
```
Ditto here for the except

calvinp0 commented on devtools/install_rmg.sh:
```diff
@@ -9,34 +9,31 @@ echo ">>> Starting RMG-Py installer..."
         -h|--help)
 cat <<EOF
-Usage: $0 [--source] [--pip] [--conda] [--rms] [--ssh]
+Usage: $0 [--source] [--pip] [--conda] [--ssh]
```
Since we have removed RMG, we should mention in the docs for user to search out the proper installation of RMG

LeenFahoum reviewed (commented):
I checked out the branch and tested the Linear TS Adapter using the showcase notebook with several unimolecular reactions. Overall, this is a very impressive feature, and the generated TS guesses worked well for most of the cases I tested.
I noticed one problematic constrained ring/puckering case in the Intra Diels-Alder monocyclic family:
`C=C1C=CC=C1 ⇌ C1=CC2CC2=C1`
I sent Alon a few related isomerization cases offline to help test and improve this behavior. My general impression is that the adapter works well overall, with this specific constrained ring case needing some additional attention.

alongd committed:
reaction: per-family recipe in get_expected_changing_bonds

Add an optional family arg so per-product_dict paths can use their own family's recipe, and stop deep-copying species just to count them in is_unimolecular/is_isomerization.

alongd committed:
docs: correct angle range claims to [-180, 180)

alongd committed:
geom_utils: add bond_order_map(), seed RNG in dihedral test

alongd committed:
zmat: raise on non-unique anchors instead of assert

Document the non-consolidated-only contract of update_zmat_by_xyz and add negative anchors tests for xyz_to_zmat.

alongd committed:
linear_utils: dedup helpers and fix stale docstrings

Reuse _connected_components/bfs_path/two_sphere_intersection/ bond_order_map instead of inline copies, keep a single PAULING_DELTA
source, drop the ignored reactive_indices arg, remove the unused has_close_h_pair_on_same_parent, and mark has_inward_migrating_group_h
as a test-support detector.

alongd committed:
linear: frame safety, reaction-state restore, per-job caching

- reorder the product into the reactant frame when rxn.atom_map exists, and skip product bond-set filtering when the mol reorder fails
- snapshot/restore rxn atom_map/family/product_dicts around execute_incore so approximations don't leak into the restart YAML
- stoichiometric counts so symmetric dissociations (A <=> B + B) are classified as dissociation
- per-reaction crash isolation and a missing-coordinates guard
- memoize map_rxn across the weight sweep; cache empty wider-family scans; pass each path's own family to get_expected_changing_bonds
- fix NameError in an except handler and the always-zero success counter
- dedup repeated contraction/ring/BFS blocks into helpers
- tests: fresh fixtures and per-test temp dirs (xdist-safe)

alongd committed:
tests: cover admission, family cache, perception hint

- scheduler unknown-family TSG admission predicate
- ReactionFamily per-process instance cache
- map_rxn None-fragment-map guard
- economic conformer generation wiring
- radical perception hint multiplicity-mismatch negative case

alongd committed:
enable the linear TS adapter by default

Add 'linear' to default_incore_adapters and to the default ts_adapters.

alongd committed:
tests: clean project dirs via addClassCleanup

tearDownClass is skipped when setUpClass raises, leaking arc/testing project dirs; register cleanups before adapter construction instead.

alongd committed:
ssh: default get_last_modified_time second path to None
